epitator/infection_annotator.py

```diff
@@ -0,0 +1,328 @@
+                # This loop groups consecutive indices into sub-lists.
+                groups = []
+                for k, g in groupby(enumerate(lax_quant_idx), lambda ix: ix[0] - ix[1]):
+                    groups.append(list(map(itemgetter(1), list(g))))
+                if len(groups) == 1:
+                    count_text = "".join([tokens[i].text for i in groups[0]])
+                    metadata["count"] = parse_count_text(count_text)
+                    debug_attributes.extend(["joined consecutive tokens", "lax count identification"])
+        except ValueError as e:
+            metadata = {}
+    if debug:
+        metadata["debug_attributes"] = debug_attributes
+    return(metadata)
+
+
+"""
+Initialized by passing in an AnnoSpan, this class will return a new
+span based on that AnnoSpan, but with 'attributes' and 'count'
+metadata slots populated, if appropriate.
+
+If the noun chunk contains a word related to infection, we include it and
+stop looking, because we assume that the noun chunk refers to a
+resultative of an infection event.
+
+If the noun chunk contains a lemma indicating a person, we continue
+looking for words in the subtree and ancestors which would indicate that
+this person was the victim of an infection event.
+
+Regardless, at the end of that process, we have a list of spaCy words and
+a list of attribute dicts. These are combined to generate the text span
+and a merged metadata object, which are used to create a AnnoSpan and
+returned.
+
+TODO: Have an argument flag for "compatibility mode", which would replace
+all attributes named "infection" with "case".
+
+If we continue down this path, I'd want to write a class which could take
+attribute-associated AnnoSpans -- say AttribSpans -- as arguments to event
+slots.
+"""
+
+
+# TODO: PUT IN INFECTION ANNOTATOR
+def has_trigger_lemmas(metadata, lemmas=["infection", "death", "hospitalization"]):
+    return any([lemma in metadata["attributes"] for lemma in lemmas])
+
+
+def has_single_count(metadata):
+    return "count" in metadata.keys() and not isinstance(metadata["count"], list)
+
+
+def from_noun_chunks_with_infection_lemmas(doc, debug=False):
+    if 'spacy.tokens' not in doc.tiers:
+        doc.add_tiers(SpacyAnnotator())
+    nc_tier = doc.tiers["spacy.noun_chunks"]
+    tokens_tier = doc.tiers["spacy.tokens"]
+
+    infection_spans = []
+
+    for nc in nc_tier:
+        # First, we check for trigger attributes in the noun chunks.
+        debug_attributes = []
+        nc_tokens = [t for t in tokens_tier.spans_contained_by_span(nc)]
+        out_tokens = nc_tokens
+        metadata = merge_dicts([
+            generate_attributes(out_tokens),
+            generate_counts(out_tokens, strict_only=False)
+        ], unique=True)
+        if has_trigger_lemmas(metadata):
+            debug_attributes.append("attributes from noun chunk")
+
+        # If the noun chunk is the subject of the root verb, we check the
```

Comment:
This comment is a under indented. Also, it could have more detail on why the code is doing that.